blog/fast-portfolio-sorts/index.qmd

```diff
@@ -0,0 +1,178 @@
+
+```{r}
+sort_dplyr <- function() {
+  crsp_monthly_dplyr |> 
+    mutate(
+      portfolio = assign_portfolio(
+        pick(everything()), "mktcap_lag", n_portfolios = 10),
+      by = "month"
+    ) |> 
+    group_by(portfolio) |> 
+    summarize(ret = mean(ret_excess, na.rm = TRUE)) |> 
+    as.data.frame()
+}
+sort_dplyr()
+```
+
+The equivalent approach in `data.table` looks as follows. Note that I deliberately don't use any pipe or intermediate assignments as to avoid any performance overhead that these might introduce. 
+
+```{r}
+sort_dt <- function() {
+  as.data.frame(crsp_monthly_dt[
+    , `:=`(portfolio = assign_portfolio(.SD, "mktcap_lag", n_portfolios = 10), by = month)
+    ][, .(ret = mean(ret_excess, na.rm = TRUE)), keyby = .(portfolio)])
+}
+sort_dt()
+```
+
+Lastly, I add the `dtplyr` implementation that also takes a `data.table` as input and internally converts `dplyr` code to `data.table` syntax. Note that the final `as.data.frame()` call is used to access the results and ensure that the result format is consistent with the other approaches. 
+
+```{r}
+sort_dtplyr <- function() {
+  crsp_monthly_dtplyr |> 
+    mutate(
+      portfolio = assign_portfolio(
+        pick(everything()), "mktcap_lag", n_portfolios = 10),
+      by = "month"
+    )  |> 
+    group_by(portfolio) |> 
+    summarize(ret = mean(ret_excess, na.rm = TRUE)) |> 
+    as.data.frame()
+}
+sort_dtplyr()
+```
+
+Now that we have verified that all code chunks create the same average excess returns per portfolio, we can proceed to the performance evaluation. 
+
+## Benchmarking results
+
+The `bench` package is a great utility for benchmarking and timing expressions in R. It provides functions that allow you to measure the execution time of  expressions or code chunks. This can be useful for comparing the performance of different approaches or implementations, or for identifying potential bottlenecks in your code. The following code evaluates each approach from above a 100 times and collects the results. 
+
+```{r}
+#| warning: false
+#| eval: false
+set.seed(1234)
+iterations <- 100
+
+results <- bench::mark(
+  sort_dplyr(), sort_dt(), sort_dtplyr(), 
+  iterations = iterations
+)
+```
+
+```{r}
+#| echo: false 
+#| eval: true
+results <- readRDS("data/bench_results.rds") 
+```
+
+The following plot shows the distributions of execution times as violin plots. You can see that `data.table` takes the lead and is followed closely by `dbplyr`, while `dtplyr` takes the third place. 
+
+```{r}
+#| warning: false
+library(ggplot2)
+
+autoplot(results, type = "violin") +
+  labs(y = NULL, x = "Sorting approach", 
+       title = "Execution time of porfolio sorts using dplyr, data.table, and dtplyr")
+```
+
```

Comment:
do we have any idea why dtplyr fails? I mean isn't that the whole promise to have a balance between syntax and speed but in this case you have no speed advantage relative to dplyr at all?

Reply:
Maybe give it a try and call lazy_dt with the data.table instead of the tibble. At least the documentation calls for "A data table (or something can can be coerced to a data table)". Not sure if there is some overhead needed whenever you call dtplyr right now.

Reply:
Good catch! It consider it as a mistake that I used the tibble in the lazy_dt()